src/TodoContext.tsx

```diff
@@ -0,0 +1,131 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import React, { useMemo, useState } from 'react';
+import { Todo } from './types/Todo';
+import { FilterField } from './types/FilterField';
+import { useLocaleStorage } from './hooks/useLocalStorage';
+
+const TodoContext = React.createContext({
+  todos: [] as Todo[],
+  setTodos: (_newTodos: Todo[]) => { },
+  title: '',
+  setTitle: (_newTitle: string) => { },
+  visibleTodos: [] as Todo[],
+  setVisibleTodos: (_newVisibleTodos: Todo[]) => { },
+  filterField: FilterField.all,
+  setFilterField: (_newFilter: FilterField) => {},
+  completeAll: () => {},
+  toggleTodo: (_toggleId: number) => {},
+  deleteTodo: (_deleteId: number) => {},
+  clearCompleted: (_event: React.MouseEvent) => {},
+  handleChange: (_event: React.ChangeEvent<HTMLInputElement>) => {},
+  handleSubmit: (_event: React.FormEvent) => {},
+});
+
+type Props = {
+  children: React.ReactNode,
+};
+
+const TodoProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useLocaleStorage<Todo[] | []>('todos', []);
+  const [title, setTitle] = useState('');
+  const [visibleTodos, setVisibleTodos] = useState(todos);
+  const [filterField, setFilterField] = useState(FilterField.all);
+
+  const completeAll = () => {
+    let state = true;
+
+    if (todos.every(t => t.completed)) {
+      state = false;
+    }
+
+    const newTodos = todos.map(t => ({
+      ...t,
+      completed: state,
+    }));
+
+    setTodos(newTodos);
+  };
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+    const newtodo = {
+      id: +new Date(),
+      title: `${title}`,
+      completed: false,
+    };
+
+    setTitle('');
+    setTodos([...todos, newtodo]);
+  };
+
+  const toggleTodo = (toggleId: number) => {
+    const newTodos = todos.map(t => ({
+      ...t,
+      completed: t.id === toggleId
+        ? !t.completed
+        : t.completed,
+    }));
+
+    setTodos(newTodos);
```

Comment:
usually, for similar things when you need update some item depends on id, better to do smth like

```
setTodos(prev => prev.map(todo => todo.id === toggleId ? { ...todo, completed: !todo.completed } : todo));
```